apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -1456,6 +1565,201 @@ impl Telemetry {
+        // want to calculate it incrementally, and have the result stored in a histogram in an
+        // observability tool.
+        // The histogram receives a range of values, and for sub-ranges, calculates how
+        // many times they were received
+        // Let's say we have N different entities. To make a histogram of entity frequencies,
+        // e want each entity frequency to be sent roughly as often as every other entity
+        // frequency, as 1/N of all values sent.
+        // Unfortunately, if we send a value on each time we see an entity, high frequency
+        // entities will be overrepresented: if a single entity appears in all requests,
+        // the frequency '1' would have the highest amount of events in the histogram,
+        // even if all the other entities have much lower frequencies. In the same way,
+        // entities with lower frequencies would be underrepresented.
+        //
+        // We solve that with sampling. With `f` the frequency of an entity, and `s(f)` its
+        // sampling rate, the probability of sending `f` to the histogram would be `f*s(f)`.
+        // If we want every frequency to appear in the same proportion, then `s(f) = c/f`,
+        // with `c` a constant.
+        // We want the sampling rate to be between 0 and 1, but as `f` gets smaller, `1/f`
+        // becomes larger. So we have to set a bound. We want to know if some entities appear
+        // often enough to warrant caching. So we will ignore entities appearing in less
+        // than 1% of requests. Which means `1/f <= 100`, so `c = 1/100` and `s(f) = 0.01/f`.
+        // Now we still want to count the low frequency entities: if a few entities appear
+        // frequently, but the largest part are rare, then caching is less useful.
+        // So we will measure them separately. First we count the number of distinct entities
+        // we have seen. Then we count the number of distinct entities we see with a frequency
+        // lower than 0.01. We assume that those frequencies are stable over the time of
+        // measurement, so the calculated frequency of rare entities will get progressively lower,
+        // until it reaches the threashold, and stay under 0.01.
+        let count = self.counts.push(representation, &1);
+        self.distinct_entities.push(representation);
+
+        let frequency = count as f64 / self.requests as f64;
+
+        // ignore entities appearing 1% of the time
+        if frequency >= 0.01 {
+            let rate = 0.01 / frequency;
+
+            let distrib = Uniform::new(0.0, 1.0);
+            let sample = rng.sample(distrib);
+
+            if sample < rate {
+                ::tracing::info!(
+                    histogram.apollo.router.operations.entity.cacheable = frequency,
+                    entity_type = %key.typename,
+                    subgraph = %key.subgraph_name,
+                    vary = %key.hashed_headers,
+                );
+            }
```

Comment:
I think it would be better to keep the same metric name with different attribute like `cache_type = cacheable` or something like that